### PR TITLE
Buzzer Support for Classic BPF Instructions

### PR DESCRIPTION
**Pull request by Negroni1**

### CODE REVIEW

**Review by amlweems — COMMENTED** [review]
lgtm!

File: pkg/cbpf/BUILD [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Copyright 2023 Google LLC
```
Comment (thatjiaozi) [inline pkg/cbpf/BUILD:1]: update to 2024? :)

File: pkg/cbpf/alu_instructions.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+func newAluInstruction[T Src](oc pb.AluOperationCode, src T) *pb.Instruction {
+	var srcType pb.SrcOperand
+	var k int32
+	switch any(src).(type) {
+	case pb.Reg:
+		srcType = pb.SrcOperand_RegSrc
+		k = int32(pb.Reg_X)
+	case int32:
+		srcType = pb.SrcOperand_Immediate
+		k = int32(src)
```
Comment (thatjiaozi) [inline pkg/cbpf/alu_instructions.go:34]: could you add a default test case where we return nil if the type is neither of those? 

This shouldn't happen (famous last words). But just in case

File: pkg/cbpf/alu_instructions.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	opcode := int32(0)
+
+	// The 3 least significant bits are the instruction class.
```
Comment (thatjiaozi) [inline pkg/cbpf/alu_instructions.go:38]: Maybe add a link to the kernel documentation as a reference for the encoding format

File: pkg/cbpf/alu_instructions.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+	return &pb.Instruction{
+		Opcode: opcode,
+		Jt:     0,
+		Jf:     0,
+		K:      k,
+	}
+}
+
+func Ret[T Src](src T) *pb.Instruction {
```
Comment (thatjiaozi) [inline pkg/cbpf/alu_instructions.go:55]: Please add a comment above in the format

// Ret [Insert explanation of what this function does]

File: pkg/cbpf/alu_instructions.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	var k int32
+	switch any(src).(type) {
+	case pb.Reg:
+		srcType = pb.SrcOperand_RegSrc
+		k = int32(pb.Reg_A)
+	case int32:
+		srcType = pb.SrcOperand_Immediate
+		k = int32(src)
+	}
+	opcode := int32(0)
```
Comment (thatjiaozi) [inline pkg/cbpf/alu_instructions.go:66]: Please add some comments on what each and operation means and link to the kernel docs where it explains so

File: pkg/cbpf/alu_instructions.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+	return &pb.Instruction{
+		Opcode: opcode,
+		Jt:     0,
+		Jf:     0,
+		K:      k,
+	}
+}
+
+func Misc(reg pb.Reg) *pb.Instruction {
```
Comment (thatjiaozi) [inline pkg/cbpf/alu_instructions.go:82]: Please add documentation comments in the form // Misc [insert explanation on what this function does]

File: pkg/cbpf/alu_instructions.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	case pb.Reg_A:
+		// TAX
+		oc = int32(0x00)
+		k = int32(pb.Reg_A)
+	case pb.Reg_X:
+		// TXA
+		oc = int32(0x80)
+		k = int32(pb.Reg_X)
+	}
+	opcode := int32(0)
```
Comment (thatjiaozi) [inline pkg/cbpf/alu_instructions.go:95]: Ditto, please explain what each masking operation does and add link to the kernel docs if possible

File: pkg/cbpf/alu_instructions_test.go [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+			wantK:                int32(X),
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.testName, func(t *testing.T) {
+			instruction := tc.instruction
+			t.Logf("Running test case %s", tc.testName)
+
+			if (instruction.Opcode & 0x07) != tc.wantInstructionClass {
```
Comment (thatjiaozi) [inline pkg/cbpf/alu_instructions_test.go:301]: consider putting the & 0x07 and each other "magic number" in a variable with a name for readability purposes, that way it would be more clear what each bit wise and operation is doing

File: pkg/cbpf/constants.go [hunk 9]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2023 Google LLC
```
Comment (thatjiaozi) [inline pkg/cbpf/constants.go:1]: 2024? also for every other file added

File: pkg/cbpf/jmp_instructions.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import (
+	pb "buzzer/proto/cbpf_go_proto"
+)
+
+func newJmpInstruction(oc pb.JmpOperationCode, jmpTrue int32,
+	jmpFalse int32, fieldK int32) *pb.Instruction {
+
+	opcode := int32(0)
+
+	// The 3 least significant bits are the instruction class.
```
Comment (thatjiaozi) [inline pkg/cbpf/jmp_instructions.go:26]: add a link to the kernel docs?

File: pkg/cbpf/jmp_instructions.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cbpf
+
+import (
+	pb "buzzer/proto/cbpf_go_proto"
+)
+
+func newJmpInstruction(oc pb.JmpOperationCode, jmpTrue int32,
```
Comment (thatjiaozi) [inline pkg/cbpf/jmp_instructions.go:21]: The function can be declared as

```
func newJmpInstruction(oc pb.JmpOperationCode, jmpTrue, jmpFalse, fieldK int32) *pb.Instruction
```

File: pkg/cbpf/jmp_instructions.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+	jmpFalse int32, fieldK int32) *pb.Instruction {
+
+	opcode := int32(0)
+
+	// The 3 least significant bits are the instruction class.
+	opcode |= (int32(pb.InsClass_InsClassJmp) & 0x07)
+
+	// The fourth bit is the source operand.
+	opcode |= (int32(pb.SrcOperand_Immediate) & 0x08)
```
Comment (thatjiaozi) [inline pkg/cbpf/jmp_instructions.go:30]: Does this mean the jumps are always immediate?

File: pkg/cbpf/jmp_instructions_test.go [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+			wantJmpFalse:         2,
+			wantK:                testK,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.testName, func(t *testing.T) {
+			instruction := tc.instruction
+			t.Logf("Running test case %s", tc.testName)
+
+			if (instruction.Opcode & 0x07) != tc.wantInstructionClass {
```
Comment (thatjiaozi) [inline pkg/cbpf/jmp_instructions_test.go:90]: same as the previous test file, consider adding each magic number to a variable for better readability

File: pkg/cbpf/st_ld_instructions.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cbpf
+
+import (
+	pb "buzzer/proto/cbpf_go_proto"
+)
+
+func newStoreLoadOperation(mode pb.StLdMode, size pb.StLdSize, class pb.InsClass, jmpTrue int8, jmpFalse int8, fieldK int32) *pb.Instruction {
```
Comment (thatjiaozi) [inline pkg/cbpf/st_ld_instructions.go:21]: Please simplify this function in the same way I mentioned for the jmp operation function

File: pkg/cbpf/st_ld_instructions_test.go [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+			wantJmpFalse: 2,
+			wantK:        testK,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.testName, func(t *testing.T) {
+			instruction := tc.instruction
+			t.Logf("Running test case %s", tc.testName)
+
+			if (instruction.Opcode & 0x07) != tc.wantClass {
```
Comment (thatjiaozi) [inline pkg/cbpf/st_ld_instructions_test.go:200]: same comment for the constants. Would it be helpful to declare them in constants.go?

**Review by thatjiaozi — CHANGES_REQUESTED** [review]
First round

File: pkg/cbpf/jmp_instructions.go (hunk 12 above)
Comment (Negroni1) [reply pkg/cbpf/jmp_instructions.go:30]: No, the implementation is incorrect. I'll add support for registers

File: pkg/cbpf/alu_instructions.go [hunk 16]
```diff
@@ -32,7 +28,13 @@ func newAluInstruction[T Src](oc pb.AluOperationCode, src T) *pb.Instruction {
 	case int32:
 		srcType = pb.SrcOperand_Immediate
 		k = int32(src)
+	default:
+		return nil
 	}
+	/*
```
Comment (thatjiaozi) [inline pkg/cbpf/alu_instructions.go:34]: (here and in the rest of the changes) The style we have been following elsewhere is to use single line comments (//). Please use those here too

**Review by thatjiaozi — APPROVED** [review]
Good job!
